### DISCUSSION TITLE
introduce specific expressions in the inital population

### DISCUSSION THREAD

**Discussion by AKHCE** [MilesCranmer/PySR#633] [Q&A]
Hi,
I am wondering if we can tell pysr that include specific expressions in the inital population. E.g. assume : I know that three expressions `y=x1**2+1`, `y=x**1.5+2`, and `y=exp(x1/10)/5` could kinda be close to the final solution, and I am interested that pysr includes these specific epxressions in the initial population. Is there any built-in feature in pysr to address such a request? 

Any advice is much appreciated.

**Comment by MilesCranmer** [comment]
See:

- https://github.com/MilesCranmer/PySR/discussions/418
- https://github.com/MilesCranmer/PySR/discussions/443

**Comment by riccardocappi** [comment]
Hi everyone, I'm new to PySR and I'm also interested in introducing specific expressions into the initial population. I saw the previous discussion on this topic, but I’m hoping for an easier way to do it or any new insights. Any guidance would be appreciated.

**Comment by MilesCranmer** [comment]
@riccardocappi now there is sort of an easier way to do it via template expressions:

```julia
expression_spec = TemplateExpressionSpec(
    expressions=["f", "g"],
    parameters={"a": 1},
    variable_names=["x"],
    combine="""
        y1 = (x + f(x))^2 + 1 + a[1]
        y2 = exp((x + f(x))/10)/(5 + a[1])
        g(x, y1, y2)
    """
)
model = PySRRegressor(expression_spec=expression_spec, **kwargs)
```

Basically this is creating a hierarchical expression where `f` and `g` are learned expressions, and `a[1]` is a learnable constant.

So this model could choose to have `f = 0.0; g = #2; a = [0.0]`, in which case the result would be equal to `y=x**2+1`. OR, it could learn `f = 0.0; g = #3; a = [0.0]` in which case the result would be equal to `y = x**1.5 + 2`. 

But it could also learn something like `f = #1; g = #2; a = [1.5]` which is equivalent to y = (2*x)**2 + 2.5`. 

it could also learn something entirely different like `f = 0.0; g = cos(#1); a = [0.0]` which would just be `y = cos(x)`.

But you can see how it is "easier" for the model to find something like @AKHCE's original suggested expressions; while still allowing the model to learn something arbitrary.

**Reply by riccardocappi** [reply]
Thank you for the quick reply!! 

So, let's assume that I have to learn the ground truth function `f = 1.5 * sin(x0 - x1)`. I have obtained from another optimization procedure (not with PySR) the function `f_init = 1.2 * sin(1.01 * x0 - 1.02 * x1) + 0.03 * x0`. My goal is to provide the latter expression as initial condition to PySR to speed-up the fitting (and hopefully obtain a more precise result).

According to your answer, I can replace the coefficients of `f_init` with some parameters `a[1], a[2], ...` and let the model learn the correct ones through the TemplateExpressionSpec method. However, i'm wondering if with this method the model is able to get rid of `0.03 * x0`, which appears in `f_init` but not in the ground truth function.
